pkg/trace/trace.go

```diff
@@ -23,19 +24,58 @@ import (
 
 var Logger = log.New()
 
-func Begin(msg string) (string, string, time.Time) {
-	pc, _, _, _ := runtime.Caller(1)
+// trace object used to grab run-time state
+type tr struct {
+	msg      string
+	funcName string
+	lineNo   int
+
+	startTime time.Time
+}
+
+func (t *tr) delta() time.Duration {
+	return time.Now().Sub(t.startTime)
+}
+
+func (t *tr) beginHdr() string {
+	return fmt.Sprintf("[BEGIN] [%s:%d]", t.funcName, t.lineNo)
+}
+
+func (t *tr) endHdr() string {
+	return fmt.Sprintf("[ END ] [%s:%d]", t.funcName, t.lineNo)
+}
+
+// begin a trace from this stack frame less the skip.
+func newTrace(msg string, skip int) *tr {
+	pc, _, line, ok := runtime.Caller(skip)
+	if !ok {
+		return nil
+	}
+
 	name := runtime.FuncForPC(pc).Name()
 
+	return &tr{
+		msg:       msg,
+		funcName:  name,
+		lineNo:    line,
+		startTime: time.Now(),
+	}
+}
+
+// Begin starts the trace.  Msg is the msg to log.
+func Begin(msg string) *tr {
+	t := newTrace(msg, 2)
+
 	if msg == "" {
-		Logger.Debugf("[BEGIN] [%s]", name)
+		Logger.Debugf(t.beginHdr())
 	} else {
-		Logger.Debugf("[BEGIN] [%s] %s", name, msg)
+		Logger.Debugf("%s %s", t.beginHdr(), t.msg)
 	}
-	return msg, name, time.Now()
+
+	return t
 }
 
-func End(msg string, name string, startTime time.Time) {
-	endTime := time.Now()
-	Logger.Debugf("[ END ] [%s] [%s] %s", name, endTime.Sub(startTime), msg)
+// End ends the trace.
+func End(t *tr) {
```

Comment:
Hmm, exported function with unexported param type.  Maybe we could rename `tr` to `Message` (trace.Message), even if it doesn't export any fields.